### PR TITLE
ci(mypy): add mypy check and adjust code for types

### LINKED ISSUE #437: Add `mypy` type check linting
Type hints can oftentimes provide helpful static analysis and developer understanding for project maintenance. We could add `mypy` as a pre-commit check to help ensure changes reach a certain standard.

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This PR adds a check through GitHub Actions for mypy in order to make sure Python types align with certain standards within pycytominer. I've also made some adjustments to the code in order to pass mypy checks.

Closes #437 

## What is the nature of your change?

- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

----
📚 Documentation preview 📚: https://pycytominer--439.org.readthedocs.build/en/439/

### CODE REVIEW

File: pycytominer/cyto_utils/output.py
Comment (gwaybio) [inline pycytominer/cyto_utils/output.py:84]: This feels dangerous - perhaps we could throw a warning that None is not supported?

File: tests/test_cyto_utils/test_output.py
Comment (gwaybio) [inline tests/test_cyto_utils/test_output.py:59]: nice!

**Review by gwaybio — COMMENTED** [review]
LGTM, but I trust Ken more to give final approval

**Review by kenibrewer — COMMENTED** [review]
This is a surprisingly low number of changes for enabling `mypy` in a repo. This might be because of the `ignore_missing_imports` and `allow_redefinitions` config options. What was the change burden like for turning those options off?

File: pycytominer/cyto_utils/output.py
Comment (gwaybio) [reply pycytominer/cyto_utils/output.py:84]: ah, I should clarify what I mean by "dangerous" - if a user explicitly sets `output_type=None` and expects certain behavior (e.g., outputting to variable and not writing), they may be surprised that we overwrite it to `output_type="csv"`. Perhaps the solution isn't to throw a warning (maybe add to docstring)? Or, we could elect to do nothing if this is standard expected behavior.

Comment (d33bs) [reply pycytominer/cyto_utils/output.py:84]: Thanks for these comments! I had to make this change in order to integrate `aggregate` with `output` properly given its default parameter types. I recognize we could change the defaults one direction or the other (with `aggregate` or `output`) but I didn't want to move too much outside of the scope of the changes here in order to get a first pass with `mypy`. I wasn't sure about rationalizing a default value for `aggregate`'s `output_type` if it wasn't going to use output. At the same time, I didn't want to change too much of `output`'s default capabilities due to how it integrates with many other functions (could get complex for the scope of this PR).

My feeling here was that if `output` was used at all that we'd expect to be "outputting" something, and that if None were passed to it for `output_type` (especially for nested function calls) that it should revert to the default string value for the parameter in `output`.

**Comment by d33bs** [comment]
Thanks @gwaybio and @kenibrewer for the reviews. 

> This is a surprisingly low number of changes for enabling mypy in a repo. 

I'm not sure, but I wonder if this has to do with the low amount of type hinting alongside compatibility with inferred types. Could be useful to make a more focused effort towards adding type hints throughout.

> What was the change burden like for turning those options off?

When I looked here I noticed we no longer needed the `allow_redefinitions` config option so I've turned this off. `ignore_missing_imports` results in the following errors / warnings. Generally I turned this on as I felt there might be many dependencies we would have to add and I wasn't certain of the benefit. Maybe this could be an iterative process where we move towards that in another change? I can also see why we might do it here.

```text
pycytominer/__config__.py:5: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/output.py:7: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/features.py:6: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/cell_locations.py:6: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/cell_locations.py:10: error: Skipping analyzing "boto3": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/cyto_utils/cell_locations.py:11: error: Skipping analyzing "botocore": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/cyto_utils/cell_locations.py:13: error: Skipping analyzing "sqlalchemy": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/cyto_utils/cell_locations_cmd.py:2: error: Skipping analyzing "fire": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/operations/transform.py:9: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/operations/transform.py:10: error: Skipping analyzing "scipy.stats": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/operations/transform.py:11: error: Skipping analyzing "sklearn.base": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/operations/transform.py:12: error: Skipping analyzing "sklearn.preprocessing": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/cyto_utils/util.py:8: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/load.py:5: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/modz.py:2: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/collate.py:75: error: Skipping analyzing "cytominer_database.ingest": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/cyto_utils/collate.py:75: error: Skipping analyzing "cytominer_database": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/cyto_utils/collate.py:76: error: Skipping analyzing "cytominer_database.munge": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/cyto_utils/write_gct.py:12: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/DeepProfiler_processing.py:7: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/DeepProfiler_processing.py:7: note: Hint: "python3 -m pip install pandas-stubs"
pycytominer/cyto_utils/DeepProfiler_processing.py:7: note: (or run "mypy --install-types" to install all missing stub packages)
pycytominer/cyto_utils/DeepProfiler_processing.py:7: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
pycytominer/aggregate.py:8: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/normalize.py:5: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/normalize.py:6: error: Skipping analyzing "sklearn.preprocessing": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pycytominer/annotate.py:7: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/cells.py:4: error: Library stubs not installed for "pandas"  [import-untyped]
pycytominer/cyto_utils/cells.py:21: error: Skipping analyzing "sqlalchemy": module is installed, but missing library stubs or py.typed marker  [import-untyped]
Found 26 errors in 16 files (checked 31 source files)
```

**Comment by d33bs** [comment]
This is now ready for another look when there's a moment @gwaybio and @kenibrewer . I've added more changes to include type stubs where reasonable. Many packages used by `pycytominer` do not have their own stubs provided within the package or through external stub packages - these are now specifically ignored via an override setting within the `pyproject.toml`. I've added additional type stub dependencies to the `dev` group, as these would now be required to successfully develop and test during development.

Generally I found that it'd be healthy to explore alternative type checkers in the future, especially `pyright`, where Microsoft has done extensive work to provide custom stubs that likely integrate via [`python-type-stubs`](https://github.com/microsoft/python-type-stubs), or [`pylyzer`](https://mtshiba.github.io/pylyzer/) which would grant speed and possibly other benefits. I didn't feel I could rationalize importing many of the relevant stubs into this project locally - these could be fairly burdensome to manage in addition to their likelihood of changing.

**Review by gwaybio — APPROVED** [review]
LGTM - let's wait on merging this until we merge #448 (after we merge 448, I am guessing you will need to resolve conflicts and rerun)

**Review by kenibrewer — APPROVED** [review]
LGTM too! Thanks for looking into the extra imports

**Comment by gwaybio** [comment]
Ready to merge @d33bs !

**Comment by d33bs** [comment]
Thanks @gwaybio and @kenibrewer for your reviews! I double checked from `main` and it looks like `mypy` tests still pass. Merging this in.

**Pull request merged**
